### 🧠 Problem with OLS when Data or Features are Large

OLS (Ordinary Least Squares) estimates coefficients using:

$$
\hat{\beta} = (X^TX)^{-1}X^Ty
$$

It minimizes the cost function:

$$
J(\beta) = \frac{1}{2m}\sum_{i=1}^{m}(y_i - X_i\beta)^2
$$

---

### ⚠️ Problems when Data or Columns Increase

1. **High Computational Cost:**  
   The matrix inversion \( (X^TX)^{-1} \) has a time complexity of \( O(n^3) \).  
   When the number of features \( n \) is large, OLS becomes very slow.

2. **High Memory Usage:**  
   \( X^TX \) is an \( n \times n \) matrix.  
   Large \( n \) means huge memory consumption.

3. **Multicollinearity:**  
   If columns (features) are correlated, \( X^TX \) becomes nearly singular.  
   Its inverse \( (X^TX)^{-1} \) may not exist or gives unstable results.

4. **Overfitting:**  
   With too many features compared to observations,  
   OLS can fit noise instead of the actual pattern.

---

✅ **Hence:**  
For large datasets or high-dimensional data,  
methods like **Gradient Descent** are preferred over pure OLS.


### ⚙️ Gradient Descent for Multiple Linear Regression

**Multiple Linear Regression** predicts the target variable \( y \) using multiple input features:

$$
\hat{y} = \theta_0 + \theta_1x_1 + \theta_2x_2 + \dots + \theta_nx_n
$$

or in vector form:

$$
\hat{y} = X\theta
$$

where:  
- \( X \) → feature matrix of size \( (m \times n) \)  
- \( \theta \) → parameter vector (weights)  
- \( m \) → number of samples  
- \( n \) → number of features  

---

### 📉 Cost Function

The goal is to minimize the **Mean Squared Error (MSE)**:

$$
J(\theta) = \frac{1}{2m} \sum_{i=1}^{m} (y_i - \hat{y_i})^2
$$

or in matrix form:

$$
J(\theta) = \frac{1}{2m}(y - X\theta)^T(y - X\theta)
$$

---

### 🔁 Gradient Descent Update Rule

At each iteration, the parameters are updated as:

$$
\theta := \theta - \alpha \frac{\partial J}{\partial \theta}
$$

where the gradient of the cost function is:

$$
\frac{\partial J}{\partial \theta} = \frac{-2}{m} X^T(y - X\theta)
$$

So the update rule becomes:

$$
\theta := \theta - \frac{\alpha}{m} X^T(X\theta - y)
$$

and the bias term (intercept) is updated as:

$$
b := b - \alpha \frac{-2}{m}\sum(y - X\theta)
$$

---

### ⚡ Advantages over OLS

1. **Efficient for large feature sets:**  
   Does not require computing \( (X^TX)^{-1} \), which is expensive when \( n \) is large.

2. **Scalable and memory-friendly:**  
   Works iteratively — suitable for high-dimensional data and large datasets.

3. **Flexible:**  
   Works for **Batch**, **Mini-batch**, and **Stochastic** variants.

---

### ⚠️ Limitations

1. Requires **feature scaling** for faster convergence.  
2. Choosing a proper **learning rate (\( \alpha \))** is critical:
   - Too high → Divergence  
   - Too low → Slow learning  
3. Needs multiple iterations to converge compared to direct OLS.

---

### ✅ Summary

Gradient Descent in **Multiple Linear Regression** optimizes all coefficients simultaneously by iteratively reducing the cost function.  
It’s the preferred method when the dataset has **many features** or **is too large** for OLS to handle efficiently.


In [152]:
import pandas as pd
import numpy as np
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.model_selection import train_test_split

In [146]:
data = pd.read_csv("Student_Performance.csv")

In [147]:
data.head()

Hours Studied  ...  Performance Index
0              7  ...               91.0
1              4  ...               65.0
2              8  ...               45.0
3              5  ...               36.0
4              7  ...               66.0

[5 rows x 6 columns]

In [148]:
data['Extracurricular Activities'] = data['Extracurricular Activities'].apply(lambda x: 1 if x == 'Yes' else 0)

In [149]:
X = data.iloc[:, 1:5].values
Y = data['Performance Index'].values

In [150]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)


In [151]:
class GradientDescent:
    def __init__(self):
        self.coef = 0
        self.intercept = 0
    def fit(self, x_train, y_train, learning_rate = 0.0001, epoches = 10000):
        self.coef = np.ones(x_train.shape[1])
        
        for i in range(epoches):
            y_pred = np.dot(x_train, self.coef) + self.intercept
            
            new_b = (-2 / len(x_train)) * np.sum(y_train - y_pred)
            self.intercept = self.intercept - (learning_rate * new_b)
            
            new_m = (-2 / len(x_train)) * (np.dot((y_train - y_pred), x_train))
            self.coef = self.coef - (learning_rate * new_m)    
            
    def predict(self, x_test):
        return np.dot(x_test, self.coef) + self.intercept
g = GradientDescent()
g.fit(x_train, y_train)
r2_score(y_test, g.predict(x_test))

0.8200137849144006

In [162]:
pd.DataFrame({"MSE":[mean_squared_error(y_test, g.predict(x_test))],
              "R2 Score":[r2_score(y_test, g.predict(x_test))]})

MSE  R2 Score
0  67.970772  0.820014

### 📊 Model Evaluation Summary

After training the model using **Batch Gradient Descent**, the results obtained are:

- **Mean Squared Error (MSE):** 67.970772  
- **R² Score:** 0.820014  

These results indicate that the model performs reasonably well —  
an \( R^2 \) score of 0.82 means that about **82% of the variance** in the target variable is explained by the model.

---

### ⚙️ Can it be Improved?

Yes, the performance **can be improved further** by:
- Increasing the number of epochs.  
- Tuning the learning rate.  
- Applying feature scaling or normalization more effectively.  
- Using regularization techniques like Ridge or Lasso Regression.

However, in this case, **the objective was not to build a high-accuracy project**,  
but to **implement and understand the Gradient Descent algorithm** for Multiple Linear Regression from scratch.

---

✅ **Conclusion:**  
The algorithm works correctly and achieves a good performance level,  
which successfully demonstrates the concept and working of **Gradient Descent** in **Multiple Linear Regression**.
